index.d.ts

```diff
@@ -465,51 +465,86 @@ export function useRead(args: unknown): any;
+export function useCache<Doc extends PathId>(
+  pathId: PathId,
+): Doc | Loading | NotFound;
 /**
- * read single value from cache
- * @param pathId PathId
- * @param field keyof Doc
- * @returns Value of key
+ * ### Read from the cache (single value of a single document)
+ * > ```ts
+ * > let myProp: Doc['prop'] = useCache<Doc, 'prop'>(
+ * >   { path: 'collection', id: 'doc-id' },
+ * >   'prop'
+ * > );
+ * > ```
+ * @param pathId PathId - { path: string; id: string }
+ * @param field - keyof Doc
+ * @returns Doc[keyof Doc] | undefined | null
  */
-function useCache<Doc extends PathId>(pathId: PathId, field: K): Doc[K];
+export function useCache<Doc extends PathId, K extends keyof Doc>(
+  pathId: PathId,
+  field: K,
+): Doc[K] | Loading | NotFound;
 /**
- * read single value from cache
- * @param pathId PathId
- * @param field keyof Doc
- * @returns Value of key
+ * ### Read from the cache (subset of a single document)
+ * > ```ts
+ * > let subset: Pick<Doc, 'prop'> = useCache<Doc, 'prop'>(
+ * > { path: 'collection', id: 'doc-id' },
+ * > ['prop']
+ * > );
+ * @param pathId { path: string; id: string; }
+ * @param field (keyof Doc)[]
+ * @returns Pick<Doc, keyof Doc> | undefined | null
  */
-function useCache<Doc extends PathId>(
+export function useCache<Doc extends PathId, K extends keyof Doc>(
   pathId: PathId,
   fields: K[],
-): Pick<Doc, K> | undefined;
+): Pick<Doc, K> | Loading | NotFound;
 /**
- * read from cache
- * @param alias string
- * @returns Alias for the query
+ * ### Read from the cache (using an alias)
+ * > ```ts
+ * > let docs: Doc[] = useCache<Doc>(someAlias);
+ * > ```
+ * @param alias - string
+ * @returns Doc[] | undefined | []
  */
-function useCache<Doc extends PathId>(alias: string): Doc[] | undefined;
+export function useCache<Doc extends PathId>(
+  alias: string,
+): Doc[] | Loading | NoResults;
 /**
- * read from cache
- * @param alias string
- * @param field field of doc
- * @returns Select keys from doc
+ * ### Read from the cache (single value of a document, using an alias)
+ * > ```ts
+ * > let props: Doc['prop'][] = useCache<Doc, 'prop'>(someAlias, 'prop');
+ * > ```
+ * @param alias - string
+ * @param field - keyof Doc
+ * @returns Doc[keyof Doc] | undefined | []
  */
-export function useCache<Doc extends PathId>(
+export function useCache<Doc extends PathId, K extends keyof Doc>(
   alias: string,
   field: K,
-): Doc[K][] | undefined;
+): Doc[K] | Loading | NoResults;
 /**
- * read from cache
+ * ### Read from the cache (a subset of a document, using an alias)
+ * > ```ts
+ * > let subsets: Pick<Doc, 'prop'>[] = useRead<Doc, 'prop'>(
+ * >   someAlias,
+ * >   ['prop']
+ * > );
+ * > ```
  * @param alias string
  * @param fields keys of doc
  * @returns Select keys from doc
  */
-export function useCache<Doc extends PathId>(
+export function useCache<Doc extends PathId, K extends keyof Doc>(
   alias: string,
   fields: K[],
 ): Pick<Doc, K>[] | undefined;
```

Comment:
For this to work it needs a function skeleton at the end. 
```ts
export function useCache(args: unknown): any;
```

Also write out the useCache function is a random ts file and make sure the type hints change properly when you change inputs and generics.

Reply:
Awesome, just tried it out using your scratchpad and they show up. Here's a screen recording of my hovering over them on VSCode: https://recordit.co/63hlv9OjD3

Reply:
Also, not sure why build is failing. Should I try and fix it before we merge this PR?